EssentialFeed/EssentialFeedTests/Feed API/LoadFeedImageCommentsFromRemoteUseCaseTests.swift

```diff
@@ -0,0 +1,225 @@
+				client.complete(withStatusCode: code, data: emptyListJSON, at: index)
+			})
+		}
+	}
+	
+	func test_load_deliversItemsOn2xxHTTPResponseWithJSONItems() {
+		let (sut, client) = makeSUT()
+		
+		let item1 = makeItem(id: UUID(),
+							 message: "a message",
+							 createdAt: (Date(timeIntervalSince1970: 1598627222), "2020-08-28T15:07:02+00:00"),
+							 username: "a username")
+		
+		let item2 = makeItem(id: UUID(),
+							 message: "another message",
+							 createdAt: (Date(timeIntervalSince1970: 1577881882), "2020-01-01T12:31:22+00:00"),
+							 username: "another username")
+		
+		let items = [item1.model, item2.model]
+		
+		let samples = [200, 201, 299]
+		
+		samples.enumerated().forEach { index, code in
+			expect(sut, toCompleteWith: .success(items), when: {
+				let json = makeItemsJSON([item1.json, item2.json])
+				client.complete(withStatusCode: code, data: json, at: index)
+			})
+		}
+	}
+	
+	func test_cancelLoadImageCommentsURLTask_cancelsClientURLRequest() {
+		let image = uniqueImage()
+		let url = URL(string: "https://a-given-url.com/image/\(image.id.uuidString)/comments")!
+		let (sut, client) = makeSUT(image: image)
+		
+		let task = sut.load() { _ in }
+		XCTAssertTrue(client.cancelledURLs.isEmpty, "Expected no cancelled URL request until task is cancelled")
+		
+		task.cancel()
+		XCTAssertEqual(client.cancelledURLs, [url], "Expected cancelled URL request after task is cancelled")
+	}
+	
+	func test_loadImageCommentsFromURL_doesNotDeliverResultAfterCancellingTask() {
+		let (sut, client) = makeSUT()
+		let nonEmptyData = Data("non-empty data".utf8)
+		
+		var received = [FeedImageCommentsLoader.Result]()
+		let task = sut.load() { received.append($0) }
+		task.cancel()
+		
+		client.complete(withStatusCode: 404, data: anyData())
+		client.complete(withStatusCode: 200, data: nonEmptyData)
+		client.complete(with: anyNSError())
+		
+		XCTAssertTrue(received.isEmpty, "Expected no received results after cancelling task")
+	}
+	
+	func test_loadImageCommentsFromURL_doesNotDeliverResultAfterSUTInstanceHasBeenDeallocated() {
+		let client = HTTPClientSpy()
+		let image = uniqueImage()
+		let url = URL(string: "https://a-given-url.com/image/\(image.id.uuidString)/comments")!
+		var sut: RemoteFeedImageCommentsLoader? = RemoteFeedImageCommentsLoader(url: url, client: client)
+		
+		var capturedResults = [FeedImageCommentsLoader.Result]()
+		_ = sut?.load { capturedResults.append($0) }
+		
+		sut = nil
+		client.complete(withStatusCode: 200, data: anyData())
+		
+		XCTAssertTrue(capturedResults.isEmpty)
+	}
+	
+	// MARK: - Helpers
+	
+	private func makeSUT(file: StaticString = #filePath, line: UInt = #line, image: FeedImage = uniqueImage()) -> (sut: RemoteFeedImageCommentsLoader, client: HTTPClientSpy) {
+		let client = HTTPClientSpy()
+		let url = URL(string: "https://a-given-url.com/image/\(image.id.uuidString)/comments")!
+		let sut = RemoteFeedImageCommentsLoader(url: url, client: client)
```

Comment:
- Remove the `image` parameter. It should be a `URL`
- It's convention to add the default file/line params *at the end* or the param list

```suggestion
	private func makeSUT(url: URL = anyURL(), file: StaticString = #filePath, line: UInt = #line) -> (sut: RemoteFeedImageCommentsLoader, client: HTTPClientSpy) {
		let client = HTTPClientSpy()
		let sut = RemoteFeedImageCommentsLoader(url: url, client: client)
```